gobblin-core/src/main/java/gobblin/source/extractor/extract/QueryBasedSource.java

```diff
@@ -12,236 +12,245 @@
 
-    // If commit policy is full and there are no failed tasks or commit
-    // policy is partial,
-    // get latest water mark as maximum of high water marks from previous
-    // tasks.
-    else {
-      if (isDataProcessedInPreviousRun) {
-        latestWaterMark = Collections.max(previousWorkUnitStateHighWatermarks);
-        LOG.info(
-            "Previous run was successful. Latest watermark - Max watermark from WorkUnitStates: " + latestWaterMark);
-      } else {
-        latestWaterMark = Collections.min(previousWorkUnitLowWatermarks);
-        LOG.info("Previous run was successful but no data found. Latest watermark - Min watermark from WorkUnitStates: "
-            + latestWaterMark);
+    for (String table : tables) {
+      try {
+        result.put(table, ConfigUtils.configToState(
+            client.getConfig(PathUtils.combinePaths(configStoreUri, QUERY_BASED_SOURCE, table).toUri())));
+      } catch (VersionDoesNotExistException | ConfigStoreFactoryDoesNotExistsException
+          | ConfigStoreCreationException e) {
+        throw new RuntimeException("Unable to get table config for " + table, e);
       }
     }
 
-    return latestWaterMark;
+    return result;
   }
 
-  private long addBackedUpSeconds(long lowWatermark, int backupSecs, String watermarkType) {
-    if (lowWatermark == ConfigurationKeys.DEFAULT_WATERMARK_VALUE) {
-      return lowWatermark;
-    }
-    WatermarkType wmType = WatermarkType.valueOf(watermarkType.toUpperCase());
-
-    switch (wmType) {
-      case SIMPLE:
-        return lowWatermark + backupSecs;
-      default:
-        Date lowWaterMarkDate = Utils.toDate(lowWatermark, "yyyyMMddHHmmss");
-        return Long
-            .parseLong(Utils.dateToString(Utils.addSecondsToDate(lowWaterMarkDate, backupSecs), "yyyyMMddHHmmss"));
+  private static SourceState getCombinedState(SourceState state, State tableSpecificState) {
+    if (tableSpecificState == null) {
+      return state;
     }
+    SourceState combinedState =
+        new SourceState(state, state.getPreviousDatasetStatesByUrns(), state.getPreviousWorkUnitStates());
+    combinedState.addAll(tableSpecificState);
+    return combinedState;
   }
 
   /**
-   * Get low water mark from the given work unit state.
+   * Pack the list of {@code WorkUnit}s into {@code MultiWorkUnit}s.
    *
-   * @param workUnitState
-   *            Work unit state
-   * @return latest low water mark
+   * TODO: this is currently a simple round-robin packing. More sophisticated bin packing may be necessary
+   * if the round-robin approach leads to mapper skew.
    */
-  private long getLowWatermarkFromWorkUnit(WorkUnitState workUnitState) {
-    String watermarkType = workUnitState
-        .getProp(ConfigurationKeys.SOURCE_QUERYBASED_WATERMARK_TYPE, ConfigurationKeys.DEFAULT_WATERMARK_TYPE);
-    long lowWaterMark = workUnitState.getWorkunit().getLowWaterMark();
+  private static List<WorkUnit> pack(List<WorkUnit> workUnits, int numOfMultiWorkunits) {
```

Comment:
It seems like bin-packing is a common pattern (Kafka, distcp-ng, here). Can we isolate the bin-packing logic and make it reusable?